### ISSUE TITLE
App\Config\Routes loaded twice

### ISSUE THREAD

**Issue by web-media** [codeigniter4/CodeIgniter4#2203]
**Describe the bug**
App\Config\Routes file is  loaded twice

**CodeIgniter 4 version**
All -Latest

**Affected class**
RouteCollection.
Codeigniter 
Codeigniter->run () ---> handleRequest() --->tryToRouteIt ()  = require APPPATH .'Config/Routes.php';

** steps to reproduce **
Add a function to App\Config\Routes file

**Comment by lonnieezell** [comment]
Just did a quick scan of the codebase and it only appears to be loaded once. `system/Config` and any environment-specific files are also loaded. 

The only way I can think of this happening otherwise is when it scans any namespaces for additional routes files it's not ignoring those two. 

I'll try to look into this in the next few days.

**Comment by lonnieezell** [comment]
Cannot reproduce. Added: 

```
function testme() { echo 'etest'; }
```

Into my routes file. Fired up the server (`php spark serve`) and loaded the page. No complaints about function already defined.

**Comment by web-media** [comment]
just tested

```php 
CRITICAL - 2019-10-09 15:10:45 --> Cannot redeclare Config\test() (previously declared in N:\a\htdocs\ci\app\Config\Routes.php:78)
#0 [internal function]: CodeIgniter\Debug\Exceptions->shutdownHandler()
#1 {main}
```

**Comment by lonnieezell** [comment]
Is this on a clean install? You don't maybe have a helper being loaded with a similarly named function do you?

**Comment by web-media** [comment]
Yes clean install . Just downloaded the latest version  from here

**Comment by lonnieezell** [comment]
@web-media what happens if you comment out the toolbar in `app/Config/Filters.php`?

**Comment by web-media** [comment]
Commented the toolbar .
 Blank page if  there is no function  in routes.php. 
 Error page if function added .
![Capture](https://user-images.githubusercontent.com/5124319/66600969-0d7ef280-eba7-11e9-8f29-afc2ecb1a1fb.PNG)

**Comment by MGatner** [comment]
Can you change it to this to get a little more helpful feedback on the what and when:
```
$routes->get('/', 'Home::index');

trace();
if (function_exists('cvvvv'))
{
	dd($routes);
}
function cvvvv() {'dsd';}

**Comment by web-media** [comment]
<ROOT>/ci/app/Config/Routes.php:76 trace()

    ⇄→<ROOT>/ci/system/CodeIgniter.php:709 require()

    ⇄→<ROOT>/ci/system/CodeIgniter.php:295 CodeIgniter\CodeIgniter->tryToRouteIt(CodeIgniter\Router\RouteCollectionInterface $routes = null)

    ⇄→<ROOT>/ci/system/CodeIgniter.php:245 CodeIgniter\CodeIgniter->handleRequest(CodeIgniter\Router\RouteCollectionInterface $routes = null, $cacheConfig, bool $returnResponse = false)

    ⇄→<ROOT>/ci/public/index.php:45 CodeIgniter\CodeIgniter->run(CodeIgniter\Router\RouteCollectionInterface $routes = null, bool $returnResponse = false)

→ Called from <ROOT>/ci/app/Config/Routes.php:76


![Capture](https://user-images.githubusercontent.com/5124319/66601355-ff7da180-eba7-11e9-993d-de255880791e.PNG)

**Comment by MGatner** [comment]
Can you run it from CLI and put the output in PasteBin (or anywhere easier to read)?

`spark routes`

**Comment by MGatner** [comment]
Initial reactions have me weirded out. I only see one Trace call in your screenshot when there should have been two, and unless I'm misunderstanding the parser I think that `dd()` should have been called before the second declaration of the function, which was not the case.

**Comment by web-media** [comment]
The only thing that came to my mind is that RouteCollection method  discoverRoutes doesn't do the job  somehow in this lines
```php
if ($file === APPPATH . 'Config/Routes.php')
				{
					continue;
				}
```
```php 
protected function discoverRoutes()
	{
		if ($this->didDiscover)
		{
			return;
		}

		// We need this var in local scope
		// so route files can access it.
		$routes = $this;

		if ($this->moduleConfig->shouldDiscover('routes'))
		{
			$files = $this->fileLocator->search('Config/Routes.php');

			foreach ($files as $file)
			{
				// Don't include our main file again...
				if ($file === APPPATH . 'Config/Routes.php')
				{
					continue;
				}

				include $file;
			}
		}

		$this->didDiscover = true;
	}
```

**Comment by lonnieezell** [comment]
Can we get some stats from you? I just tried again with a fresh clone, adding a function to `app\Config\Routes.php`, started up `php spark serve` and loaded the site. It loaded without throwing errors. So might be system-dependent thing. 

Could you provide: 
- OS type/version
- how you're serving the page (if a server other than the built-in one, please provide type and version)
- PHP version
- uri protocol (from `app\Config\App.php`)

And any other information you can think to provide.

**Comment by web-media** [comment]
XAMPP for Windows 7.3.7
Apache/2.4.39 (Win64) OpenSSL/1.1.1c PHP/7.3.7
Windows NT PC 6.1 build 7601 (Windows 7 Ultimate Edition Service Pack 1) AMD64 
PHP Version 7.3.7
 and REQUEST_URI as the protocol.

Commenting this line   //include $file;
 the routes are loaded once ... 
Seems that the bug is here 
```php
if ($file === APPPATH . 'Config/Routes.php')
				{
					continue;
				}
```
```php
protected function discoverRoutes()
	{
		if ($this->didDiscover)
		{
			return;
		}

		// We need this var in local scope
		// so route files can access it.
		$routes = $this;

		if ($this->moduleConfig->shouldDiscover('routes'))
		{
			$files = $this->fileLocator->search('Config/Routes.php');

			foreach ($files as $file)
			{
				// Don't include our main file again...
				if ($file === APPPATH . 'Config/Routes.php')
				{
					continue;
				}

				//include $file;
			}
		}

		$this->didDiscover = true;
	}
```

**Comment by web-media** [comment]
$file is
```php
N:\a\htdocs\ci\system/\Config/Routes.php
```

**Comment by lonnieezell** [comment]
Awesome. At least we can pinpoint what needs work. Thanks for sticking with it and helping us locate that!

**Comment by lonnieezell** [comment]
If you happen to create a PR with a fix, that'd be awesome. Might be a few days until I have time to really dig in, especially with needing to update my Windows install and get it working for this again.

**Comment by web-media** [comment]
you're welcome !  always glad to help . 
Not knowing yet ci4 well enough  to dig in deep  .  I'm barely scratch the surface :)

**Comment by lonnieezell** [comment]
No better way to learn :D

**Comment by lonnieezell** [comment]
@web-media Just merged a fix for that. Please pull down the latest and let me know if it works for you.

**Comment by web-media** [comment]
Solved .

**Comment by zsy78191** [comment]
I solve it by changing 
if ($file === APPPATH . 'Config/Routes.php')
to 
if (realpath($file) === APPPATH . 'Config/Routes.php')

**Comment by livelogic** [comment]
I'm using version 4.1.3 and noticed the same issue.
In my case, the solution posted by maddrid works - commenting the line in the RouteCollection module:
//include $file;
Tampering with CI system files is not my favourite approach, and i'm wondering if this is really a glitch in CI libraries or rather something else i'm missing.

**Comment by livelogic** [comment]
Still happening to me with 4.1.8
Again i can fix it with the workaround suggested from Maddrid 
`//include $file;`
CI 4.1.8, PHP 8.1.3, Windows 10, IISExpress

**Comment by kenjis** [comment]
@livelogic What value is in the `$file`?
https://github.com/codeigniter4/CodeIgniter4/blob/f969513ddedf14409bede320a1ff2f34434b9dc0/system/Router/RouteCollection.php#L362

**Comment by livelogic** [comment]
@kenjis: yes, that seem to be the culprit.

BTW, upgraded to 4.1.9 but nothing is different regarding this issue.
Two files are returned by the fileLocator->search method:

- "(project root)\app\Config\Routes.php"
- "(project root)\system\Config\Routes.php"

These paths have backslashes as directory separators, as expected in Windows environment, but the culprit line compares a path that has a Linux pattern appended:

```    if ($file === APPPATH . 'Config/Routes.php') {```

So, instead of commenting the line:

```    include $file;```

i tried this:

```    if ($file === APPPATH . 'Config\\Routes.php') {```

and things work fine (app\Config\Routes.php is loaded only once.
I haven't tried a deeper debugging but it seems clear that the paths should be normalized before comparison.

I also notice another strange thing in the same method: it loads unconditionally the system\Config\Routes.php file via include, although this file has already been loaded by the app\Config\Routes.php, via require:

```
if (file_exists(SYSTEMPATH . 'Config/Routes.php')) {
	require SYSTEMPATH . 'Config/Routes.php';
}
```
Double-loading this file doesn't seem harmful, still doesn't seem like intended behavior.

**Comment by kenjis** [comment]
@livelogic Thank you for the info.
Please try #5780

**Comment by livelogic** [comment]
@kenjis Tested the modified code in @5780, and it works fine on Windows: the Routes.php files in both app\ and system\ branches are loaded only once now.

**Comment by kenjis** [comment]
@livelogic Thank you for confirming.
The fix was merged into `develop` branch, and it will be included in v4.2.0.

**Comment by pixobit** [comment]
Is this still an issue? Ran into it on v4.4.1, and it seems to be the same behavior. If I comment out `include $file;` then it works...

**Comment by kenjis** [comment]
If you use Windows, it may be another bug. See #7930

**Comment by pixobit** [comment]
Yes, can confirm the DIRECTORY_SEPARATOR solves this issue
